src/envoy/tcp/metadata_exchange/metadata_exchange.cc

```diff
@@ -269,35 +270,55 @@ void MetadataExchangeFilter::tryReadProxyData(Buffer::Instance& data) {
       Envoy::MessageUtil::anyConvert<Envoy::ProtobufWkt::Struct>(proxy_data);
   auto key_metadata_it = value_struct.fields().find(ExchangeMetadataHeader);
   if (key_metadata_it != value_struct.fields().end()) {
-    Envoy::ProtobufWkt::Value val = key_metadata_it->second;
-    flatbuffers::FlatBufferBuilder fbb;
-    if (::Wasm::Common::extractNodeFlatBuffer(val.struct_value(), fbb)) {
-      setFilterState(config_->filter_direction_ == FilterDirection::Downstream
-                         ? DownstreamMetadataKey
-                         : UpstreamMetadataKey,
-                     absl::string_view(
-                         reinterpret_cast<const char*>(fbb.GetBufferPointer()),
-                         fbb.GetSize()));
-    }
+    updatePeer(key_metadata_it->second.struct_value());
   }
   const auto key_metadata_id_it =
       value_struct.fields().find(ExchangeMetadataHeaderId);
   if (key_metadata_id_it != value_struct.fields().end()) {
     Envoy::ProtobufWkt::Value val = key_metadata_id_it->second;
-    setFilterState(config_->filter_direction_ == FilterDirection::Downstream
-                       ? DownstreamMetadataIdKey
-                       : UpstreamMetadataIdKey,
-                   val.string_value());
+    updatePeerId(config_->filter_direction_ == FilterDirection::Downstream
+                     ? ::Wasm::Common::kDownstreamMetadataIdKey
+                     : ::Wasm::Common::kUpstreamMetadataIdKey,
+                 val.string_value());
   }
 }
 
-void MetadataExchangeFilter::setFilterState(const std::string& key,
-                                            absl::string_view value) {
+void MetadataExchangeFilter::updatePeer(
+    const Envoy::ProtobufWkt::Struct& struct_value) {
+  flatbuffers::FlatBufferBuilder fbb;
+  if (!::Wasm::Common::extractNodeFlatBuffer(struct_value, fbb)) {
+    return;
+  }
+
+  ::Envoy::Extensions::Common::Wasm::WasmStatePrototype prototype(
+      true, ::Envoy::Extensions::Common::Wasm::WasmType::FlatBuffers,
+      ::Wasm::Common::nodeInfoSchema(),
+      StreamInfo::FilterState::LifeSpan::DownstreamConnection);
+  auto state =
+      std::make_unique<::Envoy::Extensions::Common::Wasm::WasmState>(prototype);
+  state->setValue(absl::string_view(
+      reinterpret_cast<const char*>(fbb.GetBufferPointer()), fbb.GetSize()));
+
+  auto key = config_->filter_direction_ == FilterDirection::Downstream
+                 ? ::Wasm::Common::kDownstreamMetadataKey
+                 : ::Wasm::Common::kUpstreamMetadataKey;
   read_callbacks_->connection().streamInfo().filterState()->setData(
-      key,
-      std::make_unique<::Envoy::Extensions::Common::Wasm::WasmState>(value),
-      StreamInfo::FilterState::StateType::Mutable,
+      absl::StrCat("wasm.", key), std::move(state),
```

Comment:
Why attach the prefix here? Is it that Wasm runtime will attach the `wasm.` prefix by default so here it is to make the behavior consistent?

Reply:
Yes. We should really be using Wasm ABI for this, so there's a lot of boilerplate.